x-pack/plugin/snapshot-repo-test-kit/src/main/java/org/elasticsearch/repositories/blobstore/testkit/BlobAnalyzeAction.java

```diff
@@ -345,13 +350,16 @@ public StreamInput streamInput() throws IOException {
             });
         }
 
-        private void doReadBeforeWriteComplete() {
+        private void onLastReadForInitialWrite() {
             if (earlyReadNodes.isEmpty() == false) {
                 if (logger.isTraceEnabled()) {
                     logger.trace("sending read request to [{}] for [{}] before write complete", earlyReadNodes, request.getDescription());
                 }
                 readOnNodes(earlyReadNodes, true);
             }
+            if (request.getAbortWrite()) {
```

Comment:
I'm not sure this would reproduce the issue we fixed for GCP. Since we already read through all the bytes once to compute the MD5 there, we'd throw while computing the checksum and never while running the actual upload.

Without the fix this would have reproduced the issue easily but with the fix in place it seems we wouldn't catch it if e.g. say the GCP behavior was broken?

Reply:
Yes that was my understanding too, but this check does have teeth for S3 since AFAICT we only declare the length up-front and then stream the data over the wire, and then check the MD5 hash at the end (without a guarantee of cleanup if it doesn't match). The S3 API is the one that most third parties implement.

There's plans for a slightly richer `BlobStoreRepository` API focussed on repo analysis, which I think would let us implement this on GCP too, but that's future work.